pkg/agent/manager/cache/subscriber.go

```diff
@@ -24,31 +37,40 @@ type subscriber struct {
 
 type subscribers struct {
 	selMap map[string][]uuid.UUID // map of selector to UID
-	sidMap map[uuid.UUID]*subscriber
+	sidMap map[uuid.UUID]*Subscriber
 }
 
-func (s *subscribers) Add(sub *subscriber) error {
+func (s *subscribers) Add(sub *Subscriber) error {
 
-	id, err := uuid.NewV4()
-	if err != nil {
-		return err
-	}
-	s.sidMap[id] = sub
+	s.sidMap[sub.sid] = sub
 
-	selSet := selector.NewSet(sub.sel)
+	selSet := selector.NewSetFromRaw(sub.sel)
 	selPSet := selSet.Power()
 	for sel := range selPSet {
 		selStr := sel.String()
-		s.selMap[selStr] = append(s.selMap[selStr], id)
+		s.selMap[selStr] = append(s.selMap[selStr], sub.sid)
 	}
 
 	return nil
 }
 
-func (s *subscribers) Get(sels cache.Selectors) []uuid.UUID {
+func (s *subscribers) Get(sels Selectors) (subs []*Subscriber) {
+	sids := s.getSubIds(sels)
+	for _, id := range sids {
+		subs = append(subs, s.sidMap[id])
+	}
+	return
+}
+
+func (s *subscribers) Remove(sid uuid.UUID) {
+	s.sidMap[sid].done <- struct{}{}
```

Comment:
workload api handler will need to tell us when it is done with the subscriber channel... so i would think it should be writing to done channel, and we should be reading here. Also, workload api handler doesn't look like it can get subscriber id - is something else supposed to call remove?